R/layout.R

```diff
@@ -123,13 +184,14 @@ Layout <- ggproto("Layout", NULL,
     )
   },
 
-  train_ranges = function(self, coord) {
+  train_ranges = function(self) {
     compute_range <- function(ix, iy) {
       # TODO: change coord_train method to take individual x and y scales
-      coord$train(list(x = self$panel_scales$x[[ix]], y = self$panel_scales$y[[iy]]))
+      scales <- list(x = self$panel_scales$x[[ix]], y = self$panel_scales$y[[iy]])
+      self$coord$train(scales, self$coord_params)
     }
     # Switch position of all scales if CoordFlip
-    if (inherits(coord, "CoordFlip") || (inherits(coord, "CoordPolar") && coord$theta == "y")) {
+    if (inherits(self$coord, "CoordFlip") || (inherits(self$coord, "CoordPolar") && self$coord$theta == "y")) {
```

Comment:
It could be fantastic if this change/refactoring could remove the need to treat CoordFlip etc specially - that is - can we somehow create the hooks that will allow each Coord to take care of itself...